hudi-common/src/main/java/org/apache/hudi/common/table/view/IncrementalTimelineSyncFileSystemView.java

```diff
@@ -130,36 +141,50 @@ private void runIncrementalSync(HoodieTimeline timeline, TimelineDiffResult diff
     });
 
     // Now remove pending log compaction instants which were completed or removed
-    diffResult.getFinishedOrRemovedLogCompactionInstants().stream().forEach(instant -> {
+    diffResult.getFinishedOrRemovedLogCompactionInstants().stream().forEach(instantPair -> {
+      try {
+        removePendingLogCompactionInstant(instantPair.getKey(), instantPair.getValue());
+      } catch (IOException e) {
+        throw new HoodieException(e);
+      }
+    });
+
+    // Now remove pending replace instants which were completed or removed
+    diffResult.getFinishedOrRemovedReplaceInstants().stream().forEach(instantPair -> {
       try {
-        removePendingLogCompactionInstant(instant);
+        removePendingFileGroupsInPendingClustering(instantPair.getKey(), instantPair.getValue());
       } catch (IOException e) {
         throw new HoodieException(e);
       }
     });
 
-    // Add new completed instants found in the latest timeline, this also contains inflight instants.
-    diffResult.getNewlySeenInstants().stream()
-        .filter(instant -> instant.isCompleted()
-            || instant.getAction().equals(HoodieTimeline.COMPACTION_ACTION)
-            || instant.getAction().equals(HoodieTimeline.LOG_COMPACTION_ACTION))
+    HoodieTimeline newFileSystemViewTimeline =
+        newActiveTimeline.getFileSystemViewTimeline();
+    // Add new completed instants found in the latest timeline,
+    // this also contains inflight instants from rewrite timeline.
+    diffResult.getNewlySeenCompletedAndRewriteInstants()
         .forEach(instant -> {
           try {
             if (instant.getAction().equals(HoodieTimeline.COMMIT_ACTION)
                 || instant.getAction().equals(HoodieTimeline.DELTA_COMMIT_ACTION)) {
-              addCommitInstant(timeline, instant);
+              addCommitInstant(newFileSystemViewTimeline, instant);
```

Comment:
wouldn't newActiveTimeline a super set of newFileSystemViewTimeline? can we use newActiveTimeline for all from 170L to 187 ?

Reply:
It is, if I pass active timeline multiple times we will have to filter and recreate fileSystemViewTimeline. So, I created it once and passed it for methods that construct file groups. Also, all these method are private and are present in the same class so they could use a common view when building file groups.